File: sorting/quicksort.js
function swap(arr, i, j) {
    const temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp
}

function partition(arr, start, end) {
    const pivot = arr[start]
    let swapIndex = start
    for (let i = start + 1; i < end; i++) {
        if (arr[i] < pivot) {
            swapIndex++
            swap(arr, i, swapIndex)
        }
    }
    swap(arr, start, swapIndex)
    return swapIndex
}

function quicksort(arr, start=0, end=arr.length) {
    if (start < end) {
        const pivotIndex = partition(arr, start, end)
        quicksort(arr, start, pivotIndex - 1)
        quicksort(arr, pivotIndex + 1, end)    
    }
    return arr
}

module.exports = {
    partition,
    quicksort
}

File: sorting/quicksort.go
package main

// Partition ...
func Partition(arr []int, left int, right int) int {
	pivot := arr[left]
	swapIndex := left
	for i := 0; i < right; i++ {
		if arr[i] < pivot {
			swapIndex++
			arr[swapIndex], arr[i] = arr[i], arr[swapIndex]
		}
	}
	arr[left], arr[swapIndex] = arr[swapIndex], arr[left]
	return swapIndex
}

// Quicksort ...
func Quicksort(arr []int) {
	quicksortRunner(arr, 0, len(arr))
}

func quicksortRunner(arr []int, left int, right int) {
	if left < right {
		pivotIndex := Partition(arr, left, right)
		quicksortRunner(arr, left, pivotIndex-1)
		quicksortRunner(arr, pivotIndex+1, right)
	}
	return
}
